File: src/ts/jupyter-widjucks/widgets/yaml.ts
import { FnModel } from "./base";
import { safeLoad } from "js-yaml";

export class YAMLModel extends FnModel<string, any> {
  static model_name = "YAMLModel";

  defaults() {
    return {
      ...super.defaults(),
      _model_name: YAMLModel.model_name,
      value: {}
    };
  }

  theFunction(source: string) {
    return safeLoad(source);
  }
}

File: src/py/ipywidjucks/widgets/yaml.py
from .base import Base
from .base import T
from .base import W
from yaml import safe_load


@W.register
class YAML(Base):
    _model_name = T.Unicode("YAMLModel").tag(sync=True)

    source = T.Unicode("").tag(sync=True)
    value = T.Union([T.Dict(), T.List(), T.Unicode(), T.Int(), T.Float()]).tag(
        sync=True
    )

    @T.observe("source")
    def _source_changed(self, *_):
        self.value = safe_load(self.source)
